server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -1892,7 +1918,31 @@ public Integer read(StreamInput in, String key) throws IOException {
             };
 
         private ByRepo(List<Entry> entries) {
-            this.entries = List.copyOf(entries);
+            this(List.copyOf(entries), calculateStateSummaries(entries));
+        }
+
+        private ByRepo(List<Entry> entries, Tuple<Map<State, Integer>, Map<ShardState, Integer>> stateSummaries) {
+            this(entries, stateSummaries.v1(), stateSummaries.v2());
+        }
+
+        private static Tuple<Map<State, Integer>, Map<ShardState, Integer>> calculateStateSummaries(List<Entry> entries) {
```

Comment:
Hmm I think this means we do this computation on every node now which seems wasteful. Could we do it in `SnapshotsService` still, just on the master?

When I suggested doing this in `applyClusterState` I meant just updating the existing stats according to the new cluster state, not computing everything from scratch. If we have to do it from scratch every time then I guess it'd be better to happen on the stats-collection thread rather than the cluster applier. At least we could cache the results assuming they won't change before the next stats collection?